cmake/modules/FindThreads.cmake

```
@@ -0,0 +1,42 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+# Copyright (c) 2024 Google LLC.
+
+# FindThreads module for locating threads implementation.
+#
+# The module defines the following variables:
+#
+# 'Threads_FOUND'
+# Indicates if threads are supported.
+#
+# 'CMAKE_THREAD_LIBS_INIT'
+# The threads library to use. Zephyr provides threads implementation and no
+# special flags are needed, so this will be empty.
+#
+# 'CMAKE_USE_PTHREADS_INIT'
+# Indicates if threads are pthread compatible.
+#
+# This module is compatible with FindThreads module from CMake.
+# The original implementation tries to find threads library using various
+# methods (e.g. checking if pthread library is present or compiling example
+# program to check if the implementation is provided by libc), but it's not
+# able to detect pthread implementation provided by Zephyr.
+
+include(FindPackageHandleStandardArgs)
+
+set(Threads_FOUND FALSE)
+
+if ((DEFINED CONFIG_PTHREAD) OR (DEFINED CONFIG_NATIVE_APPLICATION))
```

Building targeting native_posix (NATIVE_APPLICATION) does not mean you can(should) use the host libC pthreads API. In that case you are building with the host libC but calling into that API will not produce the result you expect: when you call any of the host libC pthreads function you will relate to Linux threads instead of Zephyr threads.
```suggestion
if(DEFINED CONFIG_PTHREAD)
```

---

> when you call any of the host libC pthreads function you will relate to Linux threads instead of Zephyr threads.

That's exactly what I expect when using pthreads in native application.

Maybe the problem is that using some set of pthread functions (e.g spawning threads) in native posix can lead to unpredictable results?

Let's suppose that I have a code that uses Zephyr implementation of pthreads. How can I write unit tests for it?

---

> Let's suppose that I have a code that uses Zephyr implementation of pthreads. How can I write unit tests for it?

Depends on what you mean with unit tests. But if you mean that you want to build for a "native target", by selecting CONFIG_POSIX_API and targetting native_sim.

---

Yeah. I used `native_posix` previously, `native_sim` works for me.

I applied your suggestion. Thanks!